### DISCUSSION TITLE
How to use other functions in auth schema for RLS ?

### DISCUSSION THREAD

**Discussion by d-mok** [supabase/supabase#2086] [Questions] [answered]
I created a function `auth.email()` in the auth schema for RLS:
```
CREATE OR REPLACE FUNCTION auth.email()
 RETURNS text
 LANGUAGE sql
AS $function$
  select nullif(current_setting('request.jwt.claim.email', true), '')::text;
$function$
;
```
I then create another function `auth.is_admin()` in the auth schema that uses the `auth.email()` :
```
CREATE OR REPLACE FUNCTION auth.is_admin()
 RETURNS boolean
 LANGUAGE sql
AS $function$
  select auth.email() = 'superadmin@gmail.com'
$function$
;
```
I used `auth.is_admin()` to setup a RLS for one of my table.
But when I run my website, it returns the error:
```
permission denied for schema auth
```
It seems that when I write a function in the auth schema, I can't call other functions in the auth schema. Is it intentional by design? How can I get around it? Should I put all my function for RLS in the public schema instead? I know I can avoid calling `auth.email()` in the above simple function, but I want to be able to build complex authorization logic using functions.

**Comment by steve-chavez** [comment, marked as answer]
> permission denied for schema auth

You're missing the usage privilege on the auth schema. Try:

```sql
grant usage on schema auth to anon, authenticated;
```

**Reply by silentworks** [reply]
@steve-chavez is RLS setup on the tables in the auth schema by default? otherwise wouldn't this make it vulnerable to attacks? as in now anon can view data in those tables.

@d-mok I would probably create a new schema to handle this, I am against the idea of extending or modifying the existing supabase schemas as they could be changed at anytime by supabase themselves as part of an update.

**Reply by steve-chavez** [reply]
@silentworks No, actually anon doesn't have privileges on any table on the auth schema - you can't do a `select * from auth.users` as anon. You could say it's already restricted at the table-level, so restricting it at the row-level is not necessary.

`grant usage` would only allow anon to use the auth schema functions.

>  I would probably create a new schema to handle this, I am against the idea of extending or modifying the existing supabase schemas as they could be changed at anytime by supabase themselves as part of an update.

I agree, that sounds more sensible if you want to add your own functions. @d-mok If you do this, make sure to add the new schema on "Extra Search Path" on `/settings/api`.

**Reply by d-mok** [reply]
Thanks! But I would like to know the difference between:
1. putting my functions in the public schema
2. creating a new schema, put my functions there, and then add the new schema on "Extra Search Path" on /settings/api.
Are they just a matter of taste? Or is there any real difference in terms of security?

Also, I don't quite understand the purpose of 'Extra Search Path':
1. What does it do?
2. What happen if I don't add the new schema to it?
3. If I add the new schema to it, will the tables in that schema be exposed via the supabase API or in any way to the users?

**Reply by steve-chavez** [reply]
> putting my functions in the public schema
> creating a new schema
> Or is there any real difference in terms of security?

If you put the `is_admin` function in the `public` schema, your API users can call it through [rpc](https://supabase.io/docs/reference/javascript/rpc). If you put it in another private schema, the function will be invisible to API users.

> Also, I don't quite understand the purpose of 'Extra Search Path':
> If I add the new schema to it, will the tables in that schema be exposed via the supabase API or in any way to the users?

No, it's only for adding your schema to the search path of every request. If you don't add it, you'll get an error if you don't use the function fully qualified(schema + name) like `private.is_admin()`. More details at [db-extra-search-path](https://postgrest.org/en/stable/configuration.html#db-extra-search-path).
